### PR TITLE
Mobile: Add toolbar button reordering with up/down arrows

### PR DESCRIPTION
**Pull request by Vpatel1093**
## What this does

Adds the ability to reorder enabled toolbar buttons in the mobile editor's toolbar settings dialog. Previously, users could only toggle buttons on/off but had no control over their order. Now, each enabled button has up/down arrow controls to change its position.

## How it works

- The toolbar editor dialog (`ToolbarEditorDialog.tsx`) is split into two sections: **enabled buttons** (with up/down arrows) and **available buttons** (disabled, below a divider)
- A new `useToolbarEditorState` hook manages the reorder/toggle/reinitialize logic
- Order is persisted immediately to `editor.toolbarButtons` setting via a `useEffect` (not inside state updaters, to avoid React's "cannot update while rendering" warning)
- When the dialog reopens (e.g. after "Restore defaults"), state reinitializes from the current Redux props via a `reinitialize()` function

## Notes on button reordering

This PR uses up/down arrow buttons rather than drag-and-drop reordering. Drag-and-drop was kept out of scope to keep the change focused, but happy to add it here or as a follow-up based on maintainer preference (noting that `react-native-gesture-handler` would need to be added as a new dependency).

## Testing

Automated tests: 10 Jest tests covering partitioning, separators, move up/down (including edge cases), toggle, disabled item ordering, and reinitialize. All passing.

Manual testing performed on Android emulator (API 36):
- Reordering buttons updates the toolbar immediately on dialog close ✅
- Order persists across app restarts ✅
- Toggle + reorder interaction works correctly ✅
- Up arrow disabled on first item, down arrow disabled on last item ✅
- Restore defaults resets both the toolbar and the dialog state ✅

## Screenshots
image

### CODE REVIEW

**Comment by Vpatel1093** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by personalizedrefrigerator** [comment]
Thank you for the pull request!

Notes:
- If this is for Google Summer of Code, please review [the pull request guidelines](https://github.com/joplin/gsoc/blob/master/pull_request_guidelines.md). In particular,
   - AI-assisted contributions must follow the [AI policy](https://github.com/personalizedrefrigerator/gsoc/blob/main/ai_policy.md).
   - GSoC pull requests should be for an existing issue (usually one with the `high` or `medium` label).
   - GSoC pull requests that do not follow the pull request guidelines may be closed.
- There are a few failing tests. Please make sure that the existing mobile toolbar tests pass.
- Since this change updates the UI, screenshots and/or a screen recording in the pull request description would be helpful.
- >  Since react-native-gesture-handler is already in the project, drag-and-drop could be added without a new dependency
  
  For future reference: I don't think `react-native-gesture-handler` is currently part of the project (based on `yarn.lock` and [`yarn why -R`](https://yarnpkg.com/cli/why)). (I do see it mentioned in the "exclude" section of the expo config).
- For unit tests: In general, prefer using the actual `Setting.ts` methods to mocking them. For the necessary setup logic see, for example, [EditorToolbar.test.tsx](https://github.com/laurent22/joplin/blob/eb238efc7bb8a1fdda787196e87fb08f6b170e54/packages/app-mobile/components/EditorToolbar/EditorToolbar.test.tsx#L68).

**Comment by Vpatel1093** [comment]
@personalizedrefrigerator thanks for the feedback, do we want to add that dependency here or in a follow up PR or stick with this?

Also was looking at https://github.com/laurent22/joplin/issues/12840 and wondering if adding a "Go To Bottom" button (and possibly "Go To Top") would be good to add in a follow up? We have the CodeMirror commands already in packages/editor/CodeMirror/editorCommands/editorCommands.ts so it would be pretty trivial to implement

**Comment by mrjo118** [comment]
> @personalizedrefrigerator thanks for the feedback, do we want to add that dependency here or in a follow up PR or stick with this?
> 

Would be better to not increase the size of the PR further, given that this is your first contribution. If this change gets accepted, you can consider adding drag and drop in a followup PR. But to be honest the available button list is not that big, so I would not say its necessary at all

**Comment by personalizedrefrigerator** [comment]
> ### Testing

Since there are changes to how the list items are reported to accessibility tools, would you be willing to test this with [TalkBack](https://support.google.com/accessibility/android/answer/6283677) to make sure that the new behavior is accessible? (See [Android: Test your app's accessibility](https://developer.android.com/guide/topics/ui/accessibility/testing)) If not, I can do so, but it may take a few days.

Things to look for include:
- Whether focus is moved to a reasonable location (e.g. the next item) after checking/unchecking a list item's checkbox.
- Whether the text spoken by the screen reader after checking/unchecking an item makes sense.
- Whether the text spoken by the screen reader after moving an item up/down makes sense.
  - For example, it should be clear to the user that the item was actually moved up/down.

Thank you again for working on this!

File: packages/app-mobile/components/EditorToolbar/utils/useToolbarEditorState.test.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		const disabledNames = result.current.disabledItems.map((i: ReorderableItem) => i.commandName);
+		const boldIndex = disabledNames.indexOf('textBold');
+		const attachFileIndex = disabledNames.indexOf('attachFile');
+		const textMathIndex = disabledNames.indexOf('textMath');
+
+		// textBold should come after attachFile and before textMath in default order
+		expect(attachFileIndex).toBeLessThan(boldIndex);
+		expect(boldIndex).toBeLessThan(textMathIndex);
+	});
+});
```
Comment (laurent22) [inline packages/app-mobile/components/EditorToolbar/utils/useToolbarEditorState.test.ts:113]: I feel there are too many tests here. Please ask your AI to simplify and merge some of these tests. For example up and down arrow tests could be in the same one. Maybe some parts can be moved to functions and reuse, etc.

**Comment by laurent22** [comment]
Hello, since this pull request does not appear to be active, we are closing it to keep the PR backlog manageable. If you decide to continue working on it later, just let us know when it’s ready for review and we can reopen it.

**Comment by Vpatel1093** [comment]
> Hello, since this pull request does not appear to be active, we are closing it to keep the PR backlog manageable. If you decide to continue working on it later, just let us know when it’s ready for review and we can reopen it.

@laurent22 This is ready for review 🙂

**Comment by personalizedrefrigerator** [comment]
Comments after testing locally:
- **Auto-focus issues** (major): Thank you for adding logic to ensure that the new accessibility focus makes sense after moving an item! This logic needs adjustment to work correctly with web:
  - **Current behavior**: The button's **container** is auto-focused, rather than the button's **pressable region**. As a result, neither <kbd>space</kbd> nor <kbd>enter</kbd> activate the auto-focused button:
    
    [Screencast from 2026-03-06 14-43-28.webm](https://github.com/user-attachments/assets/efac7eb6-152d-4d79-942f-0acc09b5ab47)
    
    React Native Paper doesn't seem to expose a ref to the `TouchableRipple` used by its `IconButton`.
  - **Possible fix**: Use Joplin's `IconButton`, instead of `react-native-paper`'s `IconButton` ([example-fix.patch](https://github.com/user-attachments/files/25806042/example-patch.patch)). However, Joplin's `IconButton` attempts to display a tooltip on long-press, which doesn't seem to work properly within the toolbar editor modal.

- **Touchable sizing** (minor): The `TouchableRipple` for disabling items is shorter than the `TouchableRipple` for enabling items (22px vs 42px):
  
  | Hovering over the <code>TouchableRipple</code> checkbox | <code>TouchableRipple</code> checkbox size |
  |---|---|
  | screenshot: highlighted: touchable region|screenshot: touchable region has a height of 22px |

- **Saving state** (code structure): Currently, `useToolbarEditorState` is responsible for persisting `enabledItems`. To keep `useToolbarEditorState` focused on updating/calculating local state, consider moving the `Setting.setValue` logic to a separate hook.

**Comment by laurent22** [comment]
@Vpatel1093, please let us know if you want to keep working on this issue

**Comment by Vpatel1093** [comment]
@laurent22 yes, out for a few days but will finish

**Comment by laurent22** [comment]
Hello, I'm closing the PR for now as I'm trying to clear our backlog. Feel free to let us know if later you'd like to continue

**Comment by Vpatel1093** [comment]
@personalizedrefrigerator pushed up a commit to address your feedback - this PR can be reopened

File: packages/app-mobile/components/EditorToolbar/ToolbarEditorDialog.tsx [hunk 2]
```diff
@@ -329,7 +355,7 @@ const ToolbarEditorScreen: React.FC<EditorDialogProps> = props => {
 			<View>
 				<Text variant='bodyMedium'>{_('Check elements to display in the toolbar')}</Text>
 			</View>
-			<ScrollView style={styles.listContainer}>
+			<ScrollView style={styles.listContainer} keyboardShouldPersistTaps='handled'>
```
Comment (personalizedrefrigerator) [inline packages/app-mobile/components/EditorToolbar/ToolbarEditorDialog.tsx]: Why is `keyboardShouldPersistTaps='handled'` needed here?

**Edit**: I see that this is associated with the comment:
> Fix Space key on focused checkbox scrolling the list instead of toggling:
> add keyboardShouldPersistTaps="handled" to the ScrollView (same pattern
> as EditorToolbar)

I don't think this is what [`keyboardShouldPresistTaps`](https://reactnative.dev/docs/0.81/scrollview#keyboardshouldpersisttaps) does. For me, the scroll issue was caused by a nonpressable element having focus.

Comment (Vpatel1093) [reply packages/app-mobile/components/EditorToolbar/ToolbarEditorDialog.tsx]: Good catch - removed

**Comment by personalizedrefrigerator** [comment]
> Touchable sizing (minor): The TouchableRipple for disabling items is shorter than the TouchableRipple for enabling items (22px vs 42px):

Note: I've re-tested locally and this still seems to be an issue. The up/down arrow buttons should also be slightly larger (see [WCAG 2.2 Target Size: Minimum](https://www.w3.org/WAI/WCAG22/Understanding/target-size-minimum.html)):
screenshot: up/down arrow buttons are 20x20

(However, I can confirm that the keyboard focus issues have been fixed!)

**Review by personalizedrefrigerator — CHANGES_REQUESTED** [review]
Marking as "changes requested" as per comment above.

**Comment by Vpatel1093** [comment]
> Marking as "changes requested" as per comment above.

@personalizedrefrigerator ready for re-review!

**Comment by personalizedrefrigerator** [comment]
> @personalizedrefrigerator ready for re-review!

Thanks for the reminder! I should have this re-reviewed in the next day or two!

**Comment by personalizedrefrigerator** [comment]
I've tested this on iOS (an iOS 26 simulator) and web. The new UI seems to work well with TalkBack on MacOS.

A few notes:
- It's difficult to tell whether a "move up" or "move down" button is disabled or not. In light mode, the disabled and enabled buttons look roughly the same.
  screenshot: Move down button greyed out, but difficult to distinguish from neighboring controls

- There are a few `useEffect`s directly within React component bodies. For readability, it might (maybe) make sense to extract some of these to separate, named hooks (e.g. `useAutoFocus`).

To simplify collecting feedback on this change, I've deployed a version of this pull request to https://personalizedrefrigerator.github.io/joplin/web-client/.

**Comment by Vpatel1093** [comment]
Thanks @personalizedrefrigerator

Got those comments resolved:
image

**Comment by mrjo118** [comment]
@personalizedrefrigerator Is the requested changes status still relevant?

**Review by personalizedrefrigerator — COMMENTED** [review]
I've tested this again. I can confirm that it's easier to tell if an up/down arrow is enabled/disabled!

One possible issue: On my Android (API 29) emulator, the `TouchableRipple`s don't provide feedback (no ripple animation). However, this seems to be the case for all `TouchableRipple`s in the app, so I suspect that this issue isn't related to this pull request.

**Review by personalizedrefrigerator — APPROVED** [review]
> @personalizedrefrigerator Is the requested changes status still relevant?

I'm marking this pull request as approved.

Thank you for working on this!
